### DISCUSSION TITLE
Device in Gargepi.png image

### DISCUSSION THREAD

**Discussion by d3ra1l** [Pack3tL0ss/ConsolePi#129] [General]
What device you using in the the [garagepi.png](https://github.com/Pack3tL0ss/ConsolePi/blob/dev/readme_content/garagepi.png) picture labeled "RS232 Serial Ports Front & Back"?

**Comment by Pack3tL0ss** [comment]
It's this: https://www.pimfg.com/product-detail/BOX-EN-312

I liked that one because it allows ports front and back as an option.

There is a GPIO breakout inside, along with a FTDI 4 port dev board.  The extra UARTs on the Pi4 and dev board are all TTL level signals so they are soldered to MAX232 level-shifters (TTL <-> RS232) then to keystone jacks.

I plan to eventually create some 3D printed stuff, and share the STLs, but I have a couple of them that use a box like this or similar.

**Reply by d3ra1l** [reply]
What FTDI 4 port dev board are you using?

**Reply by Pack3tL0ss** [reply]
I have a number of them:
The one inside the setup in the Picture is this one: https://www.ebay.com/itm/174406724156
I have a number of other boards as well with varying form factor.

**Reply by n3gwg** [reply]
Wade,

I would like to build something like this. Any chance you could provide some greater guidance?
It seems that in mid-July Pi MFG shutdown for good. As such, I wonder if you ever were able to get a 3D printable type design created so it might be possible to 3D print such a box? If I were to use such a design then I am imaging it would require some sort of RS232 to TTL level converter that would get wired up to the Pi and then connected to the Keystone jacks? I saw the eBay link but that device seems to give you JTAG and other stuff but only 1 RS232 port if I am reading its specifications properly (perhaps I am not).

Frankly, I could use an electrical box and put a face plate on it, keystone plugs in it and manage to get the RS232 electronics fit in and call it a day.
https://www.homedepot.com/p/Steel-City-4-in-30-3-cu-in-Metallic-Square-Box-521711234EW-25R/202590467

or even a 16 port keystone patch panel rack insert. Put the Keystone jacks in it and then have a box (any experimenters box) to wire stuff up in too.

An alternative scenario might be to obtain a cable (not like the one substantiated in the link as follows hereupon) and simply plug it into a Pi. However, that begs another question. If you were simply going to plug in a 4 port USB serial cable such as this, is it possible to get your software running in say a VM or container and then "pass through" the USB ports to the VM or container? This of course eliminates the need for the Pi and I have several ProxMox systems running VMs anyway where I could do that.

How compatible is your project with Debian 12 (or the latest Raspian that uses Debian 12)?

4 Port cable:
http://www.ascendtech.us/copartner-usb-2-0-to-4xrs-232-db9-male_i_cbladusb4rs232n.aspx

Thanks in advance!

Stuart, N3GWG

**Reply by Pack3tL0ss** [reply]
@n3gwg Sorry for the lag, (apparently had notifications off).

The keystone jack box could be replaced without about anything. I just stumbled across the one from pi-mfg and it fit the bill, but there are a lot of similar ones out there.  I also considered using a rackmount keystone, but didn't need all the ports.

I do have a 3D model I spent hours on a while back, but put that on the back-burner  because my Prusa was getting on my nerves.  I ordered upgrades for the printer that will hopefully make it less annoying, once those finally arrive I'll try to wrap up the design.  It's a 1U rack-mount with a slide-out tray for a compute-module 4 on the official dev board.  The other half secures a USB hub (https://www.amazon.com/gp/product/B00HL7Z46K/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1) so 12 of the ports are available on the front.  And has tabs to secure a 2nd hub on the back.  This particular design doesn't have keystones, but will likely make alternatives that do (likely with trays for a Pi4/Pi5 which would leave more room.  

The FTDI USB to TTL adapter can do 4 ports of RS232, you have the option to do JTAG/RS485 etc. or you can use all 4 as RS232.  The one I linked is just one that I have, they come in a number of forms.  Search FT4232 and you'll see a lot of other options/board layouts.  You do need a TTL to RS232 level-shifter.  The most common are max232 there are other chips, but I can't recall the other one I've used.  Most of mine are max232 chips.  They come in various forms, I generally use the small solder ones.  Each max232 can provide 2 ports (2 different TTL in to 2 RS232 out).  That can be tricky on the small ones (I use stranded wire, the additional flexibility makes it less prone to snapping from the solder tab... also once tested, hot-melt glue would likely be a good idea.

ConsolePi will run on about any Linux based system, but the installer isn't customized to factor it in and make adjustments yet (it's coming, but I've got to finish the bookworm/debian 12 adjustments first).  If you expose the USB ports to the VM it should absolutely work though.  My guess is the installer would work, but would likely do some things that wouldn't apply (related to some of the other automations autohotspot, ztp, etc).  You can manually install it following the [Alternative hardware install instructions](https://github.com/Pack3tL0ss/ConsolePi?tab=readme-ov-file#alternative-hardware-installs).

As for Debian/Raspbian 12 bookworm.  consolepi-menu works, but there were a lot of changes in the daemons used to manage networking.  So autohotspot, pushbullet notifications and some other automations were impacted (noted in the known issues section).  I'm working on re-factoring that functionality around NetworkManager (the daemon that manages the network now).
